src/snowflake/snowpark/modin/plugin/_internal/index.py

```diff
@@ -82,29 +81,93 @@ def __init__(
         >>> pd.Index(list('abc'))
         Index(['a', 'b', 'c'], dtype='object')
 
+        # Snowpark pandas casts all int types to int64 by default
         >>> pd.Index([1, 2, 3], dtype="uint8")
-        Index([1, 2, 3], dtype='uint8')
+        Index([1, 2, 3], dtype='int64')
+        """
+        self.is_lazy = convert_to_lazy
+        if self.is_lazy:
+            self.set_query_compiler(
+                data=data,
+                dtype=dtype,
+                copy=copy,
+                name=name,
+                tupleize_cols=tupleize_cols,
+            )
+        else:
+            self.set_local_index(
+                data=data,
+                dtype=dtype,
+                copy=copy,
+                name=name,
+                tupleize_cols=tupleize_cols,
+            )
+
+    def set_query_compiler(
+        self,
+        # Any should be replaced with SnowflakeQueryCompiler when possible (linter won't allow it now)
+        data: ArrayLike | Any = None,
+        dtype: str | np.dtype | ExtensionDtype | None = None,
+        copy: bool = False,
+        name: object = None,
+        tupleize_cols: bool = True,
+    ) -> None:
+        """
+        Helper method to find and save query compiler when index should be lazy
         """
+        from snowflake.snowpark.modin.pandas.dataframe import DataFrame
+        from snowflake.snowpark.modin.pandas.series import Series
         from snowflake.snowpark.modin.plugin.compiler.snowflake_query_compiler import (
             SnowflakeQueryCompiler,
         )
 
-        # TODO: SNOW-1359041: Switch to lazy index implementation
-        self.is_lazy = convert_to_lazy
-        if isinstance(data, native_pd.Index):
-            self._index = data
-        elif isinstance(data, Index):
-            self._index = data.to_pandas()
+        if isinstance(data, (DataFrame, Series, Index)):
+            qc = data._query_compiler
+        elif isinstance(data, SnowflakeQueryCompiler):
+            qc = data
+        else:
+            qc = DataFrame(
+                native_pd.Index(
+                    data=data,
+                    dtype=dtype,
+                    copy=copy,
+                    name=name,
+                    tupleize_cols=tupleize_cols,
+                ).to_frame()
+            )._query_compiler
+        self._query_compiler = qc
+
+    def set_local_index(
+        self,
+        # Any should be replaced with SnowflakeQueryCompiler when possible (linter won't allow it now)
+        data: ArrayLike | Any = None,
+        dtype: str | np.dtype | ExtensionDtype | None = None,
+        copy: bool = False,
+        name: object = None,
+        tupleize_cols: bool = True,
+    ) -> None:
+        """
+        Helper method to find and save local index when index should not be lazy
```

Comment:
I wouldn't say find here since you're not searching for something. Maybe create?